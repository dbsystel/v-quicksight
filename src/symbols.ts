export const EmbeddingContextInjectionKey = Symbol(
  "The key to be used with inject when accessing the created embeddingContext for usage with the amazon-quicksight-embedding-sdk",
)
