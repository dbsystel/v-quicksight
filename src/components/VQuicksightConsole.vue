<script setup lang="ts">
import type { VQuicksightConsoleContentOptions, VQuicksightFrameOptions } from '../types'
import type {
  ConsoleContentOptions,
  ConsoleFrame,
  EmbeddingContext,
  ExperienceFrameMetadata,
  FrameOptions,
  SimpleChangeEvent,
  SimpleMessageEvent
} from 'amazon-quicksight-embedding-sdk/dist/types'
import { nanoid } from 'nanoid'
import type { Ref } from 'vue'
import { computed, inject, ref, watch } from 'vue'
import { EmbeddingContextInjectionKey } from '../symbols'

const props = withDefaults(
  defineProps<
    VQuicksightFrameOptions &
      VQuicksightConsoleContentOptions & {
        container?: string | HTMLElement // override from FrameOptions and make it optional
        id?: string
      }
  >(),
  {
    width: '100%',
    height: '100%',
    withIframePlaceholder: false
  }
)

const emit = defineEmits<{
  (e: 'change', data: { changeEvent: SimpleChangeEvent; metadata?: ExperienceFrameMetadata }): void
  (
    e: 'message',
    data: { messageEvent: SimpleMessageEvent; experienceMetadata?: ExperienceFrameMetadata }
  ): void
}>()

const embeddingContext = inject<Ref<EmbeddingContext>>(EmbeddingContextInjectionKey)

const consoleFrame = ref<ConsoleFrame>()

const containerId = computed(() => props.id || `v-quicksight-console-${nanoid(6)}`)
const frameOptions = computed<FrameOptions>(() => {
  return {
    url: props.url,
    container: props.container || `#${containerId.value}`,
    height: props.height,
    width: props.width,
    className: props.className,
    withIframePlaceholder: props.withIframePlaceholder,
    onChange: (changeEvent, metadata) => {
      emit('change', { changeEvent, metadata })
    }
  }
})
const contentOptions = computed<ConsoleContentOptions>(() => {
  return {
    locale: props.locale,
    onMessage: async (messageEvent, experienceMetadata) => {
      emit('message', { messageEvent, experienceMetadata })
    }
  }
})

async function embedConsole(ctx: EmbeddingContext) {
  consoleFrame.value = await ctx.embedConsole(frameOptions.value, contentOptions.value)
}

watch(
  () => embeddingContext?.value,
  (newValue, oldValue) => {
    if (newValue && !oldValue) {
      embedConsole(newValue)
    }
  },
  { immediate: true }
)
</script>

<template>
  <div class="v-quicksight-console" :id="containerId"></div>
</template>
