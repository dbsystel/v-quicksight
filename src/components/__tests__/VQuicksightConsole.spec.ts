import VQuicksightConsole from '../VQuicksightConsole.vue'
import { EmbeddingContextInjectionKey } from '../../symbols'
import { mount } from '@vue/test-utils'
import { describe, expect, it, vi } from 'vitest'
import { nextTick, ref } from 'vue'

const frameOptions = {
  container: '#custom-el',
  width: '300px',
  height: '700px',
  withIframePlaceholder: true,
  className: 'my-custom-class'
}
const contentOptions = {
  locale: 'en-US'
}

describe('VQuicksightConsole', () => {
  function setupComponent(props: Record<string, any> = {}) {
    const embeddingContext = {
      embedConsole: vi.fn().mockResolvedValue({})
    }

    const component = mount(VQuicksightConsole, {
      global: {
        provide: {
          [EmbeddingContextInjectionKey]: ref(embeddingContext)
        }
      },
      props: {
        ...props,
        url: 'https://my-embed-url.example.org',
        id: 'v-quicksight-console-1'
      }
    })
    return { component, embeddingContext }
  }

  it('renders as expected', () => {
    const { component } = setupComponent()
    expect(component.element).toMatchSnapshot()
  })

  it('should call embedConsole with default parameters', async () => {
    const { embeddingContext } = setupComponent()
    await nextTick()
    expect(embeddingContext.embedConsole).toHaveBeenCalledOnce()
    expect(embeddingContext.embedConsole).toHaveBeenCalledWith(
      {
        className: undefined,
        container: '#v-quicksight-console-1',
        height: '100%',
        onChange: expect.anything(),
        url: 'https://my-embed-url.example.org',
        width: '100%',
        withIframePlaceholder: false
      },
      {
        locale: undefined,
        onMessage: expect.anything()
      }
    )
  })

  it('should call embedConsole with customized parameters', () => {
    const { embeddingContext } = setupComponent({ ...frameOptions, ...contentOptions })
    expect(embeddingContext.embedConsole).toHaveBeenCalledOnce()
    expect(embeddingContext.embedConsole).toHaveBeenCalledWith(
      {
        ...frameOptions,
        onChange: expect.anything(),
        url: 'https://my-embed-url.example.org'
      },
      {
        ...contentOptions,
        onMessage: expect.anything()
      }
    )
  })
})
