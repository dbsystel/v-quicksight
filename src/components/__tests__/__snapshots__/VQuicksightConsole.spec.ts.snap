// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VQuicksightConsole > renders as expected 1`] = `
<div
  class="v-quicksight-console"
  id="v-quicksight-console-1"
/>
`;
